### PR TITLE
[move-prover][spec-simplifier] grounding work + the inlining pass

### PR DESCRIPTION
**Pull request by mengxu-fb**
Previously the simplification runs in spec-flatten, which is very
cumbersome in terms of chaining the simplification passes (due to the
fact that the `GlobalEnv` is not clonable).

The first commit relocates the transformation on the `GlobalEnv`
side into the move-model, i.e., the simplification transformations will be
co-located with the routines that build the `GlobalEnv`.

This also adds CLI support for invoking a single simplification pass
(right now the `--simplify inline` pass only) with the potential of
constructing a pipeline of passes by applying `--simplify` multiple
times.

The simplified spec should still prove. This is point checked by
```bash
mvp diem-move/diem-framework/core/sources/AccountCreationScripts.move \
    -d diem-move/diem-framework/core/sources \
    -d language/move-stdlib/sources \
    --simplify inline
```

The next commit adds the checking to the testsuite.

But that would mean running `mvp` over the test cases 3-4 times,
which is time consuming. Running verification is expensive, so merging
several functionalities for testing might make sense.

The last non-fixup commit merges `no_opaque` and `simplify`
in the test suite. These two options share many similarities that 
ake them suitable to be combined:

1. They are both move-model level modifications
2. They both focus on simplifying the spec writing or reading
3. They both involve some unrolling of code

Most of the changes in the testsuites are caused by the last commit.

## Motivation

Simplification pipeline

### Have you read the [Contributing Guidelines on pull requests](https://github.com/diem/diem/blob/main/CONTRIBUTING.md#pull-requests)?

Yes

## Test Plan

CI

### CODE REVIEW

**Review by DavidLDill — DISMISSED** [review]

File: language/move-model/src/simplifier/mod.rs [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+impl fmt::Display for SimplificationPass {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        match self {
+            Self::Inline => write!(f, "inline"),
+        }
+    }
+}
+
+impl SimplificationPass {
+    pub fn run(&self, env: &mut GlobalEnv) -> Result<()> {
```
Comment (wrwg) [inline language/move-model/src/simplifier/mod.rs:39]: Why does a simplification pass require a mut env? This is only available right after model building time, and we can't run simplification anywhere else then.

For the docgen use case of simplification, we would need both the original and the simplified spec block. At least that is how I thought about this. By default we show what the user has written, and then in an expandable block the simplified version.

Comment (mengxu-fb) [reply language/move-model/src/simplifier/mod.rs:39]: Good point. I wasn't making the distinctions clear in the code.

The code here is indeed only available right after model building time. This is for the case when I'd like to test that the simplification pass is working correctly (by proving that the spec coming out of the simplification still makes sense and verifies). For that, we need the ability to override the spec in the `GlobalEnv`, hence the `&mut` here.

For getting a simplified spec anywhere (including for the docgen), the API we are looking for is this:
`fn inline_all_exp_in_spec(env: &GlobalEnv, spec: Spec) -> Result<Spec> { .. }`, but I did not highlight this one but I should.

I'll improve the entrypoints and report back.

Comment (mengxu-fb) [reply language/move-model/src/simplifier/mod.rs]: Updated with a new commit (the last one):

[move-prover] better architecture for simplification pipeline
This creates a new trait, `SpecRewriter` that handles the
transformations on specs.

The `SpecRewriter` has three functions to be implemented by each
individual rewriter:
- `rewrite_module_spec`
- `rewrite_function_spec`
- `rewrite_inline_spec`

Multiple rewriters that implement the `SpecRewriter` trait can be
composed into a new rewriter that pipelines them. This logic is captured
in `SpecRewriterPipeline`, which also implements `SpecRewriter`.

A `SpecRewriter` has an option to write-back the newly generated spec
into a `&mut GlobalEnv` with the `override_with_rewrite` function. But
this is optional and only available in cases where a `&mut GlobalEnv`
can be obtained (e.g., during model building or right after the model
is built). If writing-back is not needed, the three `rewrite_*` funciton
should be sufficient.

**Review by wrwg — DISMISSED** [review]

**Comment by mengxu-fb** [comment]
Will land it after #9844 as there is a conflict and #9844 is more important. Probably need a re-stamping later,

**Comment by mengxu-fb** [comment]
PR rebased and ready for a restamp @wrwg @DavidLDill

**Review by junkil-park — DISMISSED** [review]

**Pull request merged**